design/open/repository-operations-error-handling.md

```diff
@@ -0,0 +1,121 @@
+
+### Cleaning Up an `initial` Repository
+A `Repository` entity with `initial` status is either being created (which will soon change the status to `active`), or already failed to create without changing status to `failed` (due to, for example, a crashed process). We will decide a `Repository` is failed if it is in `initial` state, and a sufficient time from its creation has passed. In that case, we will change the `Repository` state to `failed` and make him a candidate for deletion.
+This can be checked every time the `Repository` is accessed (trivial) and by a designated cleanup procedure (part of the above mentioned GC, or a designated scan at startup).
+ 
+##  Flows
+### CreateRepository
+* Create a `Repository` entity with its given `RepositoryID`, randomly generated `UniqueID` and `State` = `initial`, under the common `graveler` partition
+  * If failed return the error (`graveler.ErrNotUnique` will be returned in case a `Repository` entity with the same `RepositoryID` exists) - entity was not created - no harm done.
+* Generate the partition key for the new `Repository`, from its `RepositoryID` and `unique_identifier`
+* Create the initial `Commit` with a `CommitID` (same as it is done today) and store it with the partition key generated above
+  * If failed - try to set the `Repository` entry state to `failed` and return the error from the KV layer
+  * If cannot set the `Repository` state to `failed` it is still consistent, as it is still in `initial` state
+* Create the default `Branch` set to the initial `CommitID` and store it with the partition key generated above
+  * If failed - try to set the `Repository` entry state to `failed` and return the error from the KV layer
+  * If cannot set the `Repository` state to `failed` it is still consistent, as it is still in `initial` state
+* Set the `Repository` state to `active`
+  * If failed - try to set the `Repository` entry state to `failed` and return the error from the KV layer
+  * If cannot set the `Repository` state to `failed` it is still consistent, as it is still in `initial` state
+* Success
+
+### CreateBareRepository
+Just the repository is created so this is pretty trivial
+* Create a `Repository` entity with its given `RepositoryID`, randomly generated `UniqueID` and `State` = `active`, under the common `graveler` partition
+* Return the result of the creation operation. If failed to create due to existence, return `graveler.ErrNotUnique`
+
+### GetRepository
+* Get the `Repository` with the given `RepositoryID` from the common `graveler` partition
+  * If not found return `graveler.ErrRepositoryNotFound`
+  * If found and status equals `active` - return the `Repository` entity - Success
+  * Otherwise, return error (according to the state? **TBD #1**)
+
+### DeleteRepository
+* Get the `Repository`, using the `GetRepository` flow above 
+  * In case of error return the error
```

Comment:
Even if that `status` isn't `active`? Should the `Delete` flow follow the `Get` flow on these cases too?

Reply:
If the status isn't active,the first step of `GetRepository` will return an error which will stop the deletion on that spot.
Does it make sense?

Reply:
Oh, I see what you mean. Rephrased